test/client/enketo-webform.spec.js

```diff
@@ -0,0 +1,1731 @@
+         * specified type.
+         *
+         * @param {string} expected
+         */
+        const expectTypeof = ( expected ) => (
+            new ParameterPredicate( ( actual => typeof actual === expected ) )
+        );
+
+        const expectFunction = expectTypeof( 'function' );
+        const expectObject = expectTypeof( 'object' );
+
+        /**
+         * Creates a predicate to determine that a callback was provided,
+         * and call it when provided.
+         */
+        const expectCallback = new ParameterPredicate( ( callback ) => {
+            if ( typeof callback === 'function' ) {
+                callback();
+
+                return true;
+            }
+
+            return false;
+        } );
+
+        /**
+         * Creates a predicate to determine if a translator URL was provided.
+         *
+         * @param {string} expected
+         */
+        const expectLanguage = ( expected ) => (
+            new ParameterPredicate( lang => lang.includes( `/${expected}/` ) )
+        );
+
+        /**
+         * @param {object} object
+         * @param {PropertyKey} key
+         * @return {PreparedStepCache}
+         */
+        const getPreparedStep = ( object, key ) => {
+            let objectCache = preparedStepsCache.get( object );
+
+            if ( objectCache == null ) {
+                objectCache = {};
+
+                preparedStepsCache.set( object, objectCache );
+            }
+
+            let cache = objectCache[key];
+
+            if ( cache == null ) {
+                cache = {
+                    queue: [],
+                    stub: sandbox.stub( object, key ),
+                };
+
+                Object.assign( objectCache, {
+                    [key]: cache,
+                } );
+            }
+
+            return cache;
+        };
+
+        const debugLog = ( ...args ) => {
+            if ( DEBUG ) {
+                console.log( ...args );
+            }
+        };
+
+        /**
+         * Prepares a mocked initialization step which is expected to be performed.
+         * Once performed, the step is appeneded to `performedSteps` so that each
+         * step, and its order, can be verified.
+         *
+         * Behavior based on `options.stubMethod`:
+         *
+         * - 'get': the provided `options.propertyValue` is returned.
+         * - 'set': actual set value is compared to the `options.expectedValue`.
+         * - 'callsFake': actual arguments are compared to `options.expectedArgs`,
+         *   and `options.returnValue` is returned.
+         *
+         * `options.expectedArgs` items may be:
+         *
+         * - an instance of `ParameterPredicate`: its predicate will be performed
+         *   against the corresponding argument.
+         * - any other value: will be compared for deep equality.
+         *
+         * @param {InitStepOptions} options
+         * @return {InitStep}
+         */
+        const prepareInitStep = ( options ) => {
```

Comment:
As mentioned in the PR notes, I tend to prefer not writing test helpers of this sort. They tend to grow options and branchy logic over time, and tend to make tests harder to understand and maintain. In this case there's so much boilerplate involved that it felt worth the risk for now. I'm hoping the named `options` and the presence of a `description` property is helpful enough for at least making sure the usage is understandable.

Another risk is that backfilling tests for enketo-webform-edit.js and enketo-webform-view.js will likely have similar mocking boilerplate needs. There would be a temptation to extract this helper and related functionality to some shared module, further removing test logic from where the tests are.

My inclination, to mitigate these risks:

- When I move on to backfill those tests, test *all* of those modules in one file, to keep the helper as close to the tests as possible.
- Gradually move toward more directly testable initialization logic to eliminate the need for this helper altogether.